File: 139.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        boolean[] dp = new boolean[s.length()];
        for (int i = 0; i < s.length(); i++) {
            if (wordDict.contains(s.substring(0, i+1))) {
                dp[i] = true;
                continue;
            }
            for (int j = 0; j < i; j++) {
                if (!dp[j]) continue;
                if (wordDict.contains(s.substring(j+1, i+1))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[s.length()-1];
    }
}

File: 139.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp = [False] * len(s)
        for i in range(len(s)):
            for word in wordDict:
                if word == s[i-len(word)+1:i+1] and (dp[i-len(word)] or i<len(word)):
                    dp[i] = True
                    break
        return dp[-1]